allennlp/nn/util.py

```diff
@@ -567,3 +568,174 @@ def logsumexp(tensor: torch.Tensor,
+                           indices: torch.LongTensor) -> torch.Tensor:
+    """
+    The given ``indices`` of size ``(set_size, subset_size)`` specifies subsets of the ``target``
+    that each of the set_size rows should select. The `target` has size
+    ``(batch_size, sequence_length, embedding_size)``, and the resulting selected tensor has size
+    ``(batch_size, set_size, subset_size, embedding_size)``.
+
+    Parameters
+    ----------
+    target : ``torch.Tensor``, required.
+        A Tensor of shape (batch_size, sequence_length, embedding_size).
+    indices : ``torch.LongTensor``, required.
+        A LongTensor of shape (set_size, subset_size). All indices must be < sequence_length
+        as this tensor is an index into the sequence_length dimension of the target.
+
+    Returns
+    -------
+    selected : ``torch.Tensor``, required.
+        A Tensor of shape (batch_size, set_size, subset_size, embedding_size).
+    """
+    if indices.dim() != 2:
+        raise ConfigurationError("Indices passed to flatten_index_select had shape {} but "
+                                 "only 2 dimensional inputs are supported.".format(indices.size()))
+    # Shape: (batch_size, set_size * subset_size, embedding_size)
+    flattened_selected = target.index_select(1, indices.view(-1))
+
+    # Shape: (batch_size, set_size, subset_size, embedding_size)
+    selected = flattened_selected.view(target.size(0), indices.size(0), indices.size(1), -1)
+    return selected
+
+
+def get_range_vector(size: int, is_cuda: bool) -> torch.Tensor:
+    """
+    Returns a range vector with the desired size, starting at 0. The CUDA implementation
+    is meant to avoid copy data from CPU to GPU.
+    """
+    if is_cuda:
+        indices = torch.cuda.LongTensor(size).fill_(1).cumsum(0) - 1
+    else:
+        indices = torch.arange(0, size).long()
+    return Variable(indices, requires_grad=False)
+
+
+def bucket_values(distances: torch.Tensor,
+                  num_identity_buckets: int = 4,
+                  num_total_buckets: int = 10) -> torch.Tensor:
+    """
+    Places the given values (designed for distances) into ``num_total_buckets``semi-logscale
+    buckets, with ``num_identity_buckets`` of these capturing single values.
+
+    The default settings will bucket values into the following buckets:
+    [0, 1, 2, 3, 4, 5-7, 8-15, 16-31, 32-63, 64+].
+
+    Parameters
+    ----------
+    distances : ``torch.Tensor``, required.
+        A Tensor of any size, to be bucketed.
+    num_identity_buckets: int, optional (default = 4).
+        The number of identity buckets (those only holding a single value).
+    num_total_buckets : int, (default = 10)
+        The total number of buckets to bucket values into.
+
+    Returns
+    -------
+    A tensor of the same shape as the input, containing the indices of the buckets
+    the values were placed in.
+    """
+    logspace_idx = (distances.float().log()/math.log(2)).floor().long() + 3
```

Comment:
all these steps feel slightly magical, might be nice to add comments about what they're doing